servo/connectors/wavefront.py

```diff
@@ -0,0 +1,353 @@
+
+    For details on Wavefront, see the [Wavefront
+    Querying](https://docs.Wavefront.com/query_language_getting_started.html)
+    documentation.
+    """
+
+    granularity: pydantic.constr(
+        regex=r"^[smhd]"
+    ) = "m"
+
+    """The resolution of the query. The granularity resolution determines the number of data points captured across a
+    query range. A query's granularity is completely independent from any range durations specified in the WQL expression it evaluates.
+    
+    Available values: s, m, h, d.
+    """
+
+    summarization: pydantic.constr(
+        min_length=3,
+        max_length=6
+    ) = "LAST"
+    """Summarization strategy to use when bucketing points together.
+
+    Available values: MEAN, MEDIAN, MIN, MAX, SUM, COUNT, LAST, FIRST.
+    """
+
+    @property
+    def query_escaped(self) -> str:
+        # Not used as such, leaving for now in prom style
+        return re.sub(r"\{(.*?)\}", r"{{\1}}", self.query)
+
+    def __check__(self) -> servo.Check:
+        return servo.Check(
+            name=f"Check {self.name}",
+            description=f'Run Wavefront query "{self.query}"',
+        )
+
+
+class WavefrontConfiguration(servo.BaseConfiguration):
+    """WavefrontConfiguration objects describe how WavefrontConnector objects
+    capture measurements from the Wavefront metrics server.
+    """
+    api_key: pydantic.SecretStr
+    """The API key for accessing the Wavefront metrics API."""
+
+    base_url: pydantic.AnyHttpUrl = DEFAULT_BASE_URL
+    """The base URL for accessing the Wavefront metrics API.
+
+    The URL must point to the root of the Wavefront deployment. Resource paths
+    are computed as necessary for API requests.
+    """
+
+    metrics: List[WavefrontMetric]
+    """The metrics to measure from Wavefront.
+
+    Metrics must include a valid query.
+    """
+
+    @classmethod
+    def generate(cls, **kwargs) -> "WavefrontConfiguration":
+        """Generates a default configuration for capturing measurements from the
+        Wavefront metrics server.
+
+        Returns:
+            A default configuration for WavefrontConnector objects.
+        """
+        return cls(
+            description="Update the base_url and metrics to match your Wavefront configuration",
+            api_key='replace-me',
+            metrics=[
+                WavefrontMetric(
+                    "throughput",
+                    servo.Unit.REQUESTS_PER_MINUTE,
+                    query="avg(ts(appdynamics.apm.overall.calls_per_min, env=foo and app=my-app))",
+                    granularity="m",
+                    summarization="LAST",
+                ),
+                WavefrontMetric(
+                    "error_rate",
+                    servo.Unit.COUNT,
+                    query="avg(ts(appdynamics.apm.transactions.errors_per_min, env=foo and app=my-app))",
+                    granularity="m",
+                    summarization="LAST",
+                ),
```

Comment:
With the enum changes recommended above, I think this reads much more expressively:

```python
WavefrontMetric(
    "throughput",
    servo.Unit.REQUESTS_PER_MINUTE,
    query="avg(ts(appdynamics.apm.overall.calls_per_min, env=foo and app=my-app))",
    granularity=Granularity.minute,
    summarized_by=Summarization.last,
),
WavefrontMetric(
    "error_rate",
    servo.Unit.COUNT,
    query="avg(ts(appdynamics.apm.transactions.errors_per_min, env=foo and app=my-app))",
    granularity=Granularity.minute,
    summarized_by=Summarization.last,
),
```